File: backend/app/core/course_types.py
"""Canonical course types.

Owner ruling (2026-09-04): every course belongs to exactly ONE of these
three types. `course_category` and tags stay free-form and are only sorting
aids *within* a type; they can never replace or override the type.
"""

from typing import Optional, Union

COURSE_TYPES = ("meiporul", "seyappaduporul", "utporul")

DEFAULT_COURSE_TYPE = "meiporul"

COURSE_TYPE_LABELS = {
    "meiporul": "Meiporul (AR/VR)",
    "seyappaduporul": "Seyappaduporul (Skill)",
    "utporul": "Utporul (Tech)",
}


def normalize_course_type(value: Optional[str]) -> Optional[str]:
    """Normalize a client-supplied course_type to its canonical lowercase form.

    Returns None for None/blank (caller decides whether that means "leave
    unchanged" or "apply the default"). Accepts the legacy capitalized forms
    ("Meiporul", "SEYAPPADUPORUL", ...) previously stored by edit-course.
    Raises ValueError for anything outside the three allowed types so the
    routers can reject it with a clear message instead of silently storing
    a value that later shows up as an unknown filter bucket.
    """
    if value is None:
        return None
    cleaned = str(value).strip()
    if not cleaned:
        return None
    lowered = cleaned.lower()
    if lowered not in COURSE_TYPES:
        raise ValueError(
            f"course_type must be one of: {', '.join(COURSE_TYPES)}"
        )
    return lowered


# ---------------------------------------------------------------------------
# Type capability matrix (owner restructure, 2026-09-04)
# Tools/features each course type enables. The course-creation UI reads this
# (GET /api/v1/courses/type-capabilities) and shows/hides tooling; the
# backend enforces attach rules where a capability has a hard gate.
#   availability: "live" = tool exists today; "planned" = on the roadmap,
#   shown disabled (never faked).
TOOL_AVAILABILITY = {
    "video": "live", "quiz": "live", "games": "live", "h5p": "live",
    "geogebra": "live", "learning_paths": "live", "rewards": "live",
    "three_d_models": "live",      # Phase 2: GLB upload + viewer
    "live_classes": "live", "virtual_labs": "live",        # Phase 4: PhET embeds
}
TYPE_CAPABILITIES = {
    "meiporul": {  # MP — 3D/AR/VR focused
        "label": "Meiporul (AR/VR)",
        "capabilities": ["video", "quiz", "three_d_models", "geogebra", "h5p", "virtual_labs"],
    },
    "seyappaduporul": {  # SP — school students, recorded + live + games
        "label": "Seyappaduporul (School)",
        "capabilities": ["video", "quiz", "games", "h5p", "live_classes",
                         "learning_paths", "rewards", "virtual_labs"],
    },
    "utporul": {  # UP — skill oriented, everything enabled
        "label": "Utporul (Skill)",
        "capabilities": ["video", "quiz", "games", "h5p", "live_classes",
                         "learning_paths", "rewards", "geogebra",
                         "three_d_models", "virtual_labs"],
    },
}


def course_type_supports(course_type: str, capability: str) -> bool:
    cap = TYPE_CAPABILITIES.get((course_type or "").lower())
    return bool(cap and capability in cap["capabilities"])

File: frontend/src/config/courseTypes.ts
/**
 * Canonical course types — mirrors backend/app/core/course_types.py.
 * Owner ruling (2026-09-04): EVERY course belongs to exactly one of these
 * three types. Categories and tags are free-form sorting aids WITHIN a type.
 */

export interface CourseTypeOption {
  value: string
  label: string
  tamil: string
  tagline: string
  emoji: string
  cardClass: string
  selectedClass: string
}

export const COURSE_TYPES: CourseTypeOption[] = [
  {
    value: 'meiporul',
    label: 'Meiporul',
    tamil: 'மெய்ப்பொருள்',
    tagline: 'AR / VR learning',
    emoji: '🥽',
    cardClass: 'hover:border-violet-400 hover:bg-violet-50',
    selectedClass: 'border-violet-500 bg-violet-50 ring-2 ring-violet-200',
  },
  {
    value: 'seyappaduporul',
    label: 'Seyappaduporul',
    tamil: 'செயப்படுபொருள்',
    tagline: 'Skill-based training',
    emoji: '🛠️',
    cardClass: 'hover:border-emerald-400 hover:bg-emerald-50',
    selectedClass: 'border-emerald-500 bg-emerald-50 ring-2 ring-emerald-200',
  },
  {
    value: 'utporul',
    label: 'Utporul',
    tamil: 'உட்பொருள்',
    tagline: 'Tech & tools',
    emoji: '💻',
    cardClass: 'hover:border-sky-400 hover:bg-sky-50',
    selectedClass: 'border-sky-500 bg-sky-50 ring-2 ring-sky-200',
  },
]

export const COURSE_TYPE_VALUES = COURSE_TYPES.map(t => t.value)

export function courseTypeLabel(value?: string | null): string {
  if (!value) return 'Unclassified'
  const found = COURSE_TYPES.find(t => t.value === value.toLowerCase())
  return found ? `${found.label} (${found.tagline})` : value
}

export function courseTypeOption(value?: string | null): CourseTypeOption | undefined {
  if (!value) return undefined
  return COURSE_TYPES.find(t => t.value === value.toLowerCase())
}


/**
 * Tool-enablement matrix (mirrors backend app/core/course_types.py).
 * The creation wizard shows/hides tooling based on the chosen type.
 * availability "planned" renders disabled — never faked.
 */
export type CourseTool = 'virtual_labs'
  | 'video' | 'quiz' | 'games' | 'h5p' | 'geogebra'
  | 'three_d_models' | 'live_classes' | 'learning_paths' | 'rewards';

export const TOOL_LABELS: Record<CourseTool, string> = {
  virtual_labs: 'Virtual labs',
  video: 'Recorded videos',
  quiz: 'Quizzes & tests',
  games: 'Learning games',
  h5p: 'H5P interactives',
  geogebra: 'GeoGebra',
  three_d_models: '3D models (GLB)',
  live_classes: 'Live classes',
  learning_paths: 'Learning paths',
  rewards: 'Rewards & streaks',
};

export const TYPE_TOOLS: Record<string, { tools: CourseTool[]; planned: CourseTool[] }> = {
  meiporul: {
    tools: ['video', 'quiz', 'geogebra', 'h5p', 'three_d_models', 'virtual_labs', 'games', 'live_classes', 'learning_paths', 'rewards'],
    planned: [],
  },
  seyappaduporul: {
    tools: ['video', 'quiz', 'games', 'h5p', 'live_classes', 'learning_paths', 'rewards', 'virtual_labs', 'geogebra'],
    planned: ['three_d_models'],
  },
  utporul: {
    tools: ['video', 'quiz', 'games', 'h5p', 'live_classes', 'learning_paths', 'rewards', 'geogebra', 'three_d_models', 'virtual_labs'],
    planned: [],
  },
};

export function typeSupports(courseType: string | undefined, tool: CourseTool): boolean {
  if (!courseType) return true; // no type chosen yet: show nothing specialised
  const t = TYPE_TOOLS[courseType.toLowerCase()];
  return Boolean(t && t.tools.includes(tool));
}
